src/TimeSpans.jl

```diff
@@ -275,4 +310,348 @@ function time_from_index(sample_rate, sample_range::AbstractUnitRange)
+                spanstart = t
+                # 2. next step will include
+                point_will_include = false
+            else
+                push!(result, TimeSpan(spanstart, t))
+                # 2. next step will exclude
+                point_will_include = true
+            end
+        end
+    end
+
+    return TimeSpanUnion(result, is_sorted=true, may_overlap=false)
+end
+
+# set operations use a generic, higher-order function (`mergesets`) that merges
+# two AbstractTimeSpanUnion objects: the passed operator to `mergesets` should
+# indicate whether a time point should be included in the result, given the
+# membership of that point in both x and in y
+const MergableSpans = Union{TimeSpan,AbstractVector{TimeSpan}}
+function Base.intersect(x::MergableSpans, y::MergableSpans)
+    return mergesets((inx, iny) -> inx && iny, time_union(x), time_union(y))
+end
+function Base.union(x::MergableSpans, y::MergableSpans)
+    return mergesets((inx, iny) -> inx || iny, time_union(x), time_union(y))
+end
+function Base.setdiff(x::MergableSpans, y::MergableSpans)
+    return mergesets((inx, iny) -> inx && !iny, time_union(x), time_union(y))
+end
+function Base.symdiff(x::MergableSpans, y::MergableSpans)
+    return mergesets((inx, iny) -> inx ⊻ iny, time_union(x), time_union(y))
+end
+function Base.issubset(x::MergableSpans, y::MergableSpans)
+    return isempty(setdiff(x, y))
+end
+function Base.issetequal(x::MergableSpans, y::MergableSpans)
+    x, y = time_union.((x, y))
+    length(x) != length(y) && return false
+    return all(xᵢ == yᵢ for (xᵢ, yᵢ) in zip(x, y))
+end
+@static if VERSION ≥ v"1.5"
+    function Base.isdisjoint(x::MergableSpans, y::MergableSpans)
+        return isempty(intersect(x, y))
+    end
+end
+Base.in(x::TimePeriod, y::AbstractVector{TimeSpan}) = any(yᵢ -> x ∈ yᵢ, y)
+
+#####
+##### Invariant preserving operations over TimeSpanUnions 
+#####
+
+function Base.copy(x::TimeSpanUnion)
+    TimeSpanUnion(copy(x.data), is_sorted=true, may_overlap=false)
+end
+
+# some helper functions
+isneg(by::Period) = by < Nanosecond(0)
+isneg(xs) = all(x -> x < Nanosecond(0), xs)
+hasshape(by::Period) = false
+ismultidim(x) = Base.IteratorSize(x) isa Base.HasShape && length(size(x)) > 1
+
+"""
+    `shrink_all!(x::TimeSpanUnion, by)`
+
+Computes x .= extend.(x, .-by), throwing an error if it cannot be guaranteed that
+the invariants of the time span union are maintained.
+"""
+function shrink_all!(x::TimeSpanUnion, by)
+    if isneg(by)
+        throw(ArgumentError("Expected a non-negative value"))
+    elseif ismultidim(by)
+        throw(ArgumentError("Shape of time spans and `by` could produce overlapping time " *
+              "spans. Call `collect` on the time spans first."))
+    else
+        x.data .= extend.(x.data, .-by)
+    end
+
+    return x
+end
+
+"""
+    `shrink_all!(x::AbstractVector{TimeSpan}, by)`
+
+Computes x .= extend.(x, .-by)
+"""
+shrink_all!(x::AbstractVector{TimeSpan}, by) = x .= extend.(x, .-by)
```

Comment:
`shrink_all!` feels weird when you could have `shrink.(x, by)`. Additionally, why even have `shrink` when you can just use `extend.(x, -by)`. Probably would be more sensible to implement broadcasting on `TimeSpanUnion` than having these `*_all` functions.

Reply:
The (possible) advantage of these methods is that they guarantee the invariants of the returned, read-only array (sorted, with no overlap). In general, broadcasting may lead to overlapping intervals (e.g. if `by` is a `Nanosecond.(1:10)'`). The name is `shrink` because reducing the interval size always preserves the invariant, but increasing it may not.